src/algopy_testing/models/contract.py

```diff
@@ -42,24 +53,84 @@ def __init_subclass__(
+            algopy.urange | tuple[int | algopy.urange, ...] | list[int | algopy.urange] | None
         ) = None,
         state_totals: StateTotals | None = None,
     ):
         cls._name = name or cls.__name__
         cls._scratch_slots = scratch_slots
         cls._state_totals = state_totals
 
+    def _get_local_states(self) -> dict[bytes, algopy.LocalState[Any]]:
+        import algopy
+
+        local_states = {
+            attribute._key.value: attribute
+            for _, attribute in vars(self).items()
+            if isinstance(attribute, algopy.LocalState)
+        }
+
+        return local_states
+
+    def _get_global_states(self) -> dict[bytes, algopy.GlobalState[Any]]:
+        import algopy
+
+        global_states = {}
+        for key, attribute in vars(self).items():
+            if not isinstance(attribute, algopy.LocalState) and not callable(attribute):
+                if isinstance(attribute, algopy.GlobalState):
+                    global_states[attribute._key.value] = attribute
+                else:
+                    global_states[key.encode()] = attribute
+
+        return global_states
+
+    def _count_state_types(
+        self, states: dict[bytes, algopy.GlobalState[Any]] | dict[bytes, algopy.LocalState[Any]]
+    ) -> tuple[int, int]:
+        import algopy
+
+        uint_count = sum(
+            1
+            for state in states.values()
+            if is_instance(
+                (
+                    state.type_
+                    if is_instance(state, (algopy.GlobalState | algopy.LocalState))
+                    else state
+                ),
+                algopy.UInt64
+                | algopy.arc4.UIntN
+                | algopy.arc4.BigUIntN
+                | algopy.Asset
+                | algopy.Application,
+            )
+        )
+        bytes_count = len(states) - uint_count
+        return uint_count, bytes_count
+
+    def _get_global_state_totals(
+        self, global_states: dict[bytes, algopy.GlobalState[Any]] | None = None
+    ) -> tuple[int, int]:
+        return self._count_state_types(global_states or self._get_global_states())
+
+    def _get_local_state_totals(
+        self, local_states: dict[bytes, algopy.LocalState[Any]] | None = None
+    ) -> tuple[int, int]:
+        return self._count_state_types(local_states or self._get_local_states())
+
     def approval_program(self) -> algopy.UInt64 | bool:
         raise NotImplementedError("`approval_program` is not implemented.")
 
     def clear_state_program(self) -> algopy.UInt64 | bool:
         raise NotImplementedError("`clear_state_program` is not implemented.")
 
-    def __hash__(self) -> int:
-        return hash(self._name)
-
     def __getattribute__(self, name: str) -> Any:
+        from algopy_testing.context import get_test_context
+
+        context = get_test_context()
+        context.set_active_contract(self)
```

Comment:
Can we assume that a contract is active just because an attribute is being accessed? This would also happen outside of a function being called e.g. in the arrange or assert sections of a test.

Maybe we could set a contract as active via the arc4 decorators or approval / clear methods in non arc4 contracts?